### PR TITLE
Adding support for specifying a sns subscription filter policy

### PR DESCRIPTION
**Pull request by neilcampbell**

### CODE REVIEW

File: src/main/scala/seek/aws/sns/SubscribeTopic.scala [hunk 1]
```diff
@@ -34,34 +38,69 @@ class SubscribeTopic extends AwsTask {
...
+          _  <- updateSubscriptions(t, merge(ps, es)).run(c)
         } yield ()
     }
   }
 
-  private def addSubscriptions(topicArn: String, ps: List[PendingSubscription]): Kleisli[IO, AmazonSNS, Unit] =
+  private def updateSubscriptions(topicArn: String, ps: List[PendingSubscription]): Kleisli[IO, AmazonSNS, Unit] =
+    ps match {
+      case Nil    => Kleisli.liftF(IO.unit)
+      case h :: t => for {
```
Comment (ashlineldridge) [inline src/main/scala/seek/aws/sns/SubscribeTopic.scala]: For consistency with the rest of the codebase can you put multiline statements within a `for` comprehension on a newline after the arrow?

File: src/main/scala/seek/aws/sns/SubscribeTopic.scala [hunk 2]
```diff
@@ -34,34 +38,69 @@ class SubscribeTopic extends AwsTask {
...
+        _   <- maybeUpdateFilterPolicy(arn, h.filterPolicy)
+        _   <- updateSubscriptions(topicArn, t)
+      } yield ()
+    }
+
+  private def maybeSubscribeToTopic(topicArn: String, ps: PendingSubscription): Kleisli[IO, AmazonSNS, String] =
+    Kleisli { c =>
+      ps.arn match {
+        case Some(a) => IO.pure(a)
+        case None => IO(c.subscribe(topicArn, ps.protocol, ps.endpoint).getSubscriptionArn)
```
Comment (ashlineldridge) [inline src/main/scala/seek/aws/sns/SubscribeTopic.scala]: Please align `=>`.

File: src/main/scala/seek/aws/sns/SubscribeTopic.scala [hunk 3]
```diff
@@ -34,34 +38,69 @@ class SubscribeTopic extends AwsTask {
...
-        case Nil    => IO.unit
-        case h :: t =>
-          for {
-            _ <- IO(c.subscribe(topicArn, h.protocol, h.endpoint))
-            _ <- addSubscriptions(topicArn, t).run(c)
-          } yield ()
+      policy match {
+        case Some(p) => IO {
+          val r = new SetSubscriptionAttributesRequest(arn, "FilterPolicy", p)
+          c.setSubscriptionAttributes(r)
```
Comment (ashlineldridge) [inline src/main/scala/seek/aws/sns/SubscribeTopic.scala]: The `new SetSubscription...` shouldn't be inside the `IO`. Probably be simpler just to do do `c.setSubscriptionAttributes(new Set...`. Either way can you please align the `=>` by making the `None` come first?

File: src/main/scala/seek/aws/sns/SubscribeTopic.scala [hunk 4]
```diff
@@ -34,34 +38,69 @@ class SubscribeTopic extends AwsTask {
...
     }
 
-  private def uniquePending(pending: Seq[PendingSubscription], existing: Seq[Subscription]): Set[PendingSubscription] =
-    pending.toSet.diff(existing.map(s => PendingSubscription(s.getProtocol, s.getEndpoint)).toSet)
+  private def merge(pending: List[PendingSubscription], existing: List[Subscription]): List[PendingSubscription] =
+    existing.foldLeft(pending) { (z, es) =>
+      z.zipWithIndex
+        .find(psi => {
+          val (ps, _) = psi
+          ps.protocol == es.getProtocol && ps.endpoint == es.getEndpoint
```
Comment (ashlineldridge) [inline src/main/scala/seek/aws/sns/SubscribeTopic.scala]: You can simplify these by going:
```
.find { case (ps, i) => ...  } 
```

File: src/main/scala/seek/aws/sns/SubscribeTopic.scala [hunk 5]
```diff
@@ -34,34 +38,69 @@ class SubscribeTopic extends AwsTask {
...
+        })
+        .map(psi => {
```
Comment (ashlineldridge) [inline src/main/scala/seek/aws/sns/SubscribeTopic.scala]: Same here.

Comment (neilcampbell) [reply src/main/scala/seek/aws/sns/SubscribeTopic.scala]: That reads much better!

**Pull request merged**